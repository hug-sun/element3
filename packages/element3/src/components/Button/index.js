import ElButton from './src/Button.vue'

ElButton.install = function (app) {
  app.component(ElButton.name, ElButton)
}

export { ElButton }
