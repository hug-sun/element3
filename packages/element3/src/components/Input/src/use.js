import { ref, computed, watch, reactive, toRefs, inject, onMounted } from 'vue'
import calcTextareaHeight from './calcTextareaHeight'

export const useClass = (inputSize, inputExceed, props, attrs, slots) => {
  return computed(() => {
    const classes = [
      inputSize.value ? 'el-input--' + inputSize.value : '',
      props.type === 'textarea' ? 'el-textarea' : 'el-input'
    ]

    if (attrs.disabled) {
      classes.push('is-disabled')
    }

    if (inputExceed.value) {
      classes.push('is-exceed')
    }

    if (slots.prepend || slots.append) {
      classes.push('el-input-group')
    }

    if (slots.append) {
      classes.push('el-input-group--append')
    }

    if (slots.prepend) {
      classes.push('el-input-group--prepend')
    }
    if (slots.prefix || props.prefixIcon) {
      classes.push('el-input--prefix')
    }

    if (
      slots.suffix ||
      props.suffixIcon ||
      slots.suffixIcon ||
      attrs.clearable ||
      attrs.showPassword
    ) {
      classes.push('el-input--suffix')
    }

    return classes
  })
}

export const useInput = (props, cxt, textarea) => {
  const input = ref(null)

  const elFormItem = inject('elFormItem', {})

  const elFormChange = inject('elForm.change', () => {})

  const {
    modelValue,
    size,
    suffixIcon,
    clearable,
    showPassword,
    showWordLimit
  } = toRefs(props)

  const nativeInputValue = computed(() => {
    return modelValue.value === null || modelValue.value === undefined
      ? ''
      : String(modelValue.value)
  })

  const textLength = computed(() => {
    if (typeof modelValue.value === 'number') {
      return String(modelValue.value).length >= Number(cxt.attrs.maxlength)
        ? Number(cxt.attrs.maxlength)
        : String(modelValue.value).length
    }

    return modelValue.value.length >= Number(cxt.attrs.maxlength)
      ? Number(cxt.attrs.maxlength)
      : modelValue.value.length
  })

  const elFormItemSize = computed(() => {
    return elFormItem.elFormItemSize || ''
  })

  const inputSize = computed(() => {
    return size?.value || elFormItemSize.value
  })

  const inputExceed = computed(() => {
    return props.modelValue?.length >= Number(cxt.attrs.maxlength)
      ? true
      : false
  })

  const getInput = () =>
    computed({
      get: () => (input.value ? input.value : textarea.value),
      set: (value) => {
        if (input?.value) {
          input.value.value = value
        }

        if (textarea?.value) {
          textarea.value.value = value
        }
      }
    })

  const clearValue = () => {
    cxt.emit('update:modelValue', '')
  }

  const setNativeInputValue = () => {
    const input = getInput()
    if (!input) return
    input.value = nativeInputValue.value
  }

  const getSuffixVisible = computed(() => {
    return (
      cxt.slots.suffix ||
      suffixIcon?.value ||
      clearable?.value ||
      showPassword?.value ||
      showWordLimit?.value
    )
  })

  onMounted(() => {
    setNativeInputValue(nativeInputValue.value)
  })

  watch(
    () => props.modelValue,
    () => {
      setNativeInputValue()
      if (props.validateEvent) {
        elFormChange()
      }
    }
  )

  return {
    input,
    getInput,
    nativeInputValue,
    textLength,
    clearValue,
    inputSize,
    getSuffixVisible,
    inputExceed
  }
}

export const useInputEvent = (emit) => {
  const handleInput = (event) => {
    emit('update:modelValue', event.target.value)
    emit('input', event.target.value)
  }

  const handleFocus = (event) => {
    emit('focus', event)
  }

  const handleBlur = (event) => emit('blur', event)

  const handleClear = () => {
    emit('update:modelValue', '')
    emit('clear')
  }

  const onChange = (event) => {
    emit('change', event.target.value)
  }

  return {
    handleInput,
    handleFocus,
    handleBlur,
    handleClear,
    onChange
  }
}

export const useInputMethod = (element) => {
  const focus = () => {
    element.value.focus()
  }

  const blur = () => {
    element.value.blur()
  }

  const select = () => {
    element.value.select()
  }

  return {
    focus,
    blur,
    select
  }
}

export const useTextarea = (props) => {
  const textarea = ref(null)

  const { autosize, type, resize } = toRefs(props)

  const state = reactive({
    textareaCalcStyle: {}
  })

  const textareaStyle = computed(() => {
    return Object.assign({}, state.textareaCalcStyle, { resize: resize?.value })
  })

  watch(
    () => props.modelValue,
    () => resizeTextarea()
  )

  const resizeTextarea = () => {
    if (type.value !== 'textarea') return
    if (!autosize.value) {
      state.textareaCalcStyle = {
        minHeight: calcTextareaHeight(textarea.value).minHeight
      }

      return
    }
    const minRows = autosize.value.minRows
    const maxRows = autosize.value.maxRows
    state.textareaCalcStyle = calcTextareaHeight(
      textarea.value,
      minRows,
      maxRows
    )
  }

  return { textarea, textareaStyle, resizeTextarea }
}
