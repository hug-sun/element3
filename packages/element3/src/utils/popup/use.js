import { computed, onBeforeMount, onBeforeUnmount, onUnmounted } from 'vue'
import { addClass, removeClass, hasClass } from '../../utils/dom'

let zIndex = 2000

let useBodyScrollCounter = 0

function useZindex() {
  return computed(() => zIndex++)
}

function useBodyScroll({ lockScroll }) {
  const body = document.body

  useBodyScrollCounter++
  onBeforeMount(() => {
    if (lockScroll && !hasClass(body, 'el-popup-parent--hidden')) {
      addClass(body, 'el-popup-parent--hidden')
    }
  })

  onUnmounted(() => {
    useBodyScrollCounter--
    if (lockScroll || useBodyScrollCounter <= 1) {
      removeClass(document.body, 'el-popup-parent--hidden')
    }
  })
}

export { useZindex, useBodyScroll }
