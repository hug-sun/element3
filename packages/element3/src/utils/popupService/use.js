import { computed, onBeforeMount, onUnmounted, ref } from 'vue'
import { addClass, removeClass, hasClass } from '../dom'

let zIndex = 2000

let useBodyScrollCounter = 0

function useZindex() {
  return zIndex++
}

export const useStyle = () => {
  const style = ref({})

  function setStyle(options) {
    style.value = Object.assign({}, style, options)
  }
  return {
    style,
    setStyle
  }
}

function useBodyScroll({ lockScroll }) {
  const body = document.body

  onBeforeMount(() => {
    if (lockScroll) {
      useBodyScrollCounter++
      if (!hasClass(body, 'el-popup-parent--hidden')) {
        addClass(body, 'el-popup-parent--hidden')
      }
    }
  })

  onUnmounted(() => {
    if (lockScroll) {
      useBodyScrollCounter--
      if (useBodyScrollCounter <= 0) {
        removeClass(document.body, 'el-popup-parent--hidden')
      }
    }
  })
}

export { useZindex, useBodyScroll }
