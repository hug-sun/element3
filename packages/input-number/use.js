import { computed, inject, unref, getCurrentInstance, watch } from 'vue'
export const useSize = (size) => {
  const elFormItem = inject('elFormItem', {})

  const _elFormItemSize = computed(() => {
    return unref(elFormItem.elFormItemSize)
  })

  const inputNumberSize = computed(() => {
    return (
      size.value ||
      _elFormItemSize ||
      (getCurrentInstance().ctx.$ELEMENT || {}).size
    )
  })
  return inputNumberSize
}
