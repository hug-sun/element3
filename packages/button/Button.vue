<template>
  <button
    class="el-button"
    :class="classes"
    :type="nativeType"
    :disabled="buttonDisabled || loading"
  >
    <i class="el-icon-loading" v-if="loading"></i>
    <i :class="icon" v-else-if="icon"></i>
    <span v-if="$slots.default">
      <slot></slot>
    </span>
  </button>
</template>

<script>
import { useGlobalOptions } from '../../src/use/globalConfig'
import { toRefs, inject, computed } from 'vue'
export default {
  name: 'ElButton',
  props: {
    size: {
      type: String,
      validator(val) {
        if (val === '') return true
        return ['medium', 'small', 'mini'].indexOf(val) !== -1
      }
    },
    type: {
      type: String,
      validator(val) {
        return (
          ['primary', 'success', 'warning', 'danger', 'info', 'text'].indexOf(
            val
          ) !== -1
        )
      }
    },
    nativeType: {
      type: String,
      default: 'button'
    },
    plain: Boolean,
    round: Boolean,
    circle: Boolean,
    loading: Boolean,
    disabled: Boolean,
    icon: String
  },
  setup(props) {
    const { size, disabled } = toRefs(props)

    const buttonSize = useButtonSize(size)
    const buttonDisabled = useButtonDisabled(disabled)
    const classes = useClasses({
      props,
      size: buttonSize,
      disabled: buttonDisabled
    })

    return {
      buttonDisabled,
      classes
    }
  }
}

const useClasses = ({ props, size, disabled }) => {
  return computed(() => {
    return [
      size.value ? `el-button--${size.value}` : '',
      props.type ? `el-button--${props.type}` : '',
      {
        'is-plain': props.plain,
        'is-round': props.round,
        'is-circle': props.circle,
        'is-loading': props.loading,
        'is-disabled': disabled.value
      }
    ]
  })
}

const useButtonDisabled = (disabled) => {
  return computed(() => {
    const elForm = inject('elForm', {})

    return disabled?.value || elForm.disabled
  })
}

const useButtonSize = (size) => {
  const globalConfig = useGlobalOptions()
  return computed(() => {
    const elFormItem = inject('elFormItem', {})
    return size?.value || elFormItem.elFormItemSize || globalConfig.size
  })
}
</script>
