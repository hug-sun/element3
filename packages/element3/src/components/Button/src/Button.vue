<template>
  <button
    class="el-button"
    :class="classes"
    :type="nativeType"
    :disabled="buttonDisabled || loading"
  >
    <i class="el-icon-loading" v-if="loading"></i>
    <i :class="icon" v-else-if="icon"></i>
    <span v-if="$slots.default">
      <slot></slot>
    </span>
  </button>
</template>

<script lang="ts">
import { props } from './props'
import { useGlobalOptions } from '../../../use/globalConfig'
import { toRefs, inject, computed, defineComponent, Ref } from 'vue'
export default defineComponent({
  name: 'ElButton',
  props,
  setup(props) {
    const { size, disabled } = toRefs(props)
    const buttonSize = useButtonSize(size)
    const buttonDisabled = useButtonDisabled(disabled)
    const classes = useClasses({
      props,
      size: buttonSize,
      disabled: buttonDisabled
    })
    return {
      buttonDisabled,
      classes
    }
  }
})

const useClasses = ({ props, size, disabled }) => {
  return computed(() => {
    return [
      size.value ? `el-button--${size.value}` : '',
      props.type ? `el-button--${props.type}` : '',
      {
        'is-plain': props.plain,
        'is-round': props.round,
        'is-circle': props.circle,
        'is-loading': props.loading,
        'is-disabled': disabled.value
      }
    ]
  })
}

const useButtonDisabled = (disabled: Ref) => {
  return computed(() => {
    const elForm = inject('elForm', null)

    return disabled?.value || elForm?.disabled
  })
}

const useButtonSize = (size: Ref) => {
  const globalConfig = useGlobalOptions()
  return computed(() => {
    const elFormItem = inject('elFormItem', null)
    return size?.value || elFormItem?.elFormItemSize || globalConfig.size
  })
}
</script>
