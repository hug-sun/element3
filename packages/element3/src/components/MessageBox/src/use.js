import { toRefs, nextTick, unref, onMounted, onUnmounted, computed } from 'vue'
import { usePopup } from '../../../composables/popup'
import { isFunction } from '@vue/shared'
export function useHandleList(state, instance, validate) {
  const { close } = usePopup(state)
  const {
    visible,
    callback,
    category,
    inputType,
    distinguishCancelAndClose,
    closeOnClickModal,
    action,
    beforeClose
  } = toRefs(state)
  const closeHandle = () => {
    visible.value = false
    close()
    nextTick(() => {
      unref(callback)(state.action, instance.proxy)
    })
  }
  const handleAction = (_action) => {
    if (unref(category) === 'prompt' && _action === 'confirm' && !validate()) {
      return
    }
    action.value = _action
    if (isFunction(beforeClose.value)) {
      beforeClose.value(_action, instance.proxy, closeHandle)
    } else {
      closeHandle()
    }
  }

  const handleInputEnter = () => {
    if (unref(inputType) !== 'textarea') {
      return handleAction('confirm')
    }
  }

  const handleWrapperClick = () => {
    if (unref(closeOnClickModal)) {
      handleAction(unref(distinguishCancelAndClose) ? 'close' : 'cancel')
    }
  }

  return {
    closeHandle,
    handleAction,
    handleWrapperClick,
    handleInputEnter
  }
}

export function watchElement(state, handleAction, closeHandle) {
  const {
    closeOnHashChange,
    closeOnPressEscape,
    distinguishCancelAndClose
  } = toRefs(state)
  const { open } = usePopup(state)
  const handleKeyup = (element = {}) => {
    if (element.code !== 'Escape') return
    if (unref(closeOnPressEscape)) {
      handleAction(unref(distinguishCancelAndClose) ? 'close' : 'cancel')
    }
  }
  onMounted(() => {
    if (unref(closeOnHashChange)) {
      window.addEventListener('hashchange', closeHandle)
    }
    window.addEventListener('keyup', handleKeyup)
    nextTick(() => {
      open()
    })
  })
  onUnmounted(() => {
    if (unref(closeOnHashChange)) {
      window.removeEventListener('hashchange', closeHandle)
    }
    window.removeEventListener('keyup', handleKeyup)
  })
}

export function classIcon(iconClass, type) {
  return computed(() => {
    return unref(iconClass) || (unref(type) ? `el-icon-${unref(type)}` : '')
  })
}
