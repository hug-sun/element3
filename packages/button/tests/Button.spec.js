import Button from '../Button.vue'
import { mount } from '@vue/test-utils'
import { reactive } from 'vue'
import { setupGlobalOptions } from '../../../src/use/globalConfig'

describe('Button.vue', () => {
  it('snapshot', () => {
    const wrapper = mount(Button)
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should show content', () => {
    const content = 'foo'

    const wrapper = mount(Button, {
      slots: {
        default: content
      }
    })

    expect(wrapper.text()).toContain(content)
  })

  describe('set button size', () => {
    it('by props.size', () => {
      const size = 'small'

      const wrapper = mount(Button, {
        props: {
          size
        }
      })

      expect(wrapper.classes()).toContain(`el-button--${size}`)
    })

    it('by elFormItem.elFormItemSize', () => {
      const size = 'small'
      const wrapper = mount(Button, {
        props: {
          size: ''
        },
        global: {
          provide: {
            elFormItem: reactive({
              elFormItemSize: size
            })
          }
        }
      })

      expect(wrapper.classes()).toContain(`el-button--${size}`)
    })

    it('by global config ', () => {
      const size = 'small'
      const wrapper = mount(Button, {
        props: {
          size: ''
        },
        global: {
          plugins: [
            setupGlobalOptions({
              size
            })
          ]
        }
      })

      expect(wrapper.classes()).toContain(`el-button--${size}`)
    })
  })

  it('set button type by prop type ', () => {
    const type = 'success'

    const wrapper = mount(Button, {
      props: {
        type
      }
    })

    expect(wrapper.classes()).toContain(`el-button--${type}`)
  })

  it('set button plain by prop type', () => {
    const wrapper = mount(Button, {
      props: {
        plain: true
      }
    })

    expect(wrapper.classes()).toContain(`is-plain`)
  })
  it('set button round by prop type', () => {
    const wrapper = mount(Button, {
      props: {
        round: true
      }
    })

    expect(wrapper.classes()).toContain(`is-round`)
  })

  it('set button circle by prop type', () => {
    const wrapper = mount(Button, {
      props: {
        circle: true
      }
    })

    expect(wrapper.classes()).toContain(`is-circle`)
  })

  it('set button loading by prop loading', async () => {
    const wrapper = mount(Button, {
      props: {
        loading: true
      }
    })

    expect(wrapper.classes()).toContain(`is-loading`)
  })

  describe('set button disabled', () => {
    it('by props.disabled', () => {
      const wrapper = mount(Button, {
        props: {
          disabled: true
        }
      })

      expect(wrapper.classes()).toContain(`is-disabled`)
      expect(wrapper.attributes()).toHaveProperty('disabled')
    })

    it('by elForm.disabled', () => {
      const wrapper = mount(Button, {
        global: {
          provide: {
            elForm: reactive({
              disabled: true
            })
          }
        }
      })

      expect(wrapper.classes()).toContain(`is-disabled`)
      expect(wrapper.attributes()).toHaveProperty('disabled')
    })
  })

  describe('set button icon', () => {
    it(' by props.icon', () => {
      const wrapper = mount(Button, {
        props: {
          icon: 'el-icon-edit'
        }
      })

      expect(wrapper.find('.el-icon-edit').exists()).toBe(true)
    })

    it("don't show icon when loading eq true", () => {
      const wrapper = mount(Button, {
        props: {
          icon: 'el-icon-edit',
          loading: true
        }
      })

      expect(wrapper.find('.el-icon-edit').exists()).toBe(false)
      expect(wrapper.find('.el-icon-loading').exists()).toBe(true)
    })
  })

  it('set native-type by props.native-type', () => {
    const nativeType = 'reset'

    const wrapper = mount(Button, {
      props: {
        nativeType
      }
    })

    expect(wrapper.attributes('type')).toBe(nativeType)
  })
})
