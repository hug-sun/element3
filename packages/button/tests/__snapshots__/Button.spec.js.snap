// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button.vue snapshot 1`] = `
<button
  class="el-button"
  type="button"
>
  <!--v-if-->
  <!--v-if-->
</button>
`;
